keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -89,25 +90,23 @@ TBA
 
 ## Proposal
 
-We need to introduce an algorithm-agnostic HPA object configuration that will configure each particular HPA scaling behavior.
-For both direction (scale up and scale down) there should be a `Constraint` field with the following parameters:
+We need to introduce an algorithm-agnostic HPA object configuration that will allow configuration of individual HPAs.
+To customize the scaling behavior we should add a `Behavior` field with the following fields:
 
-- `periodSeconds` - a parameter to specify the time period for the constraint, in seconds
-- `percent` - a parameter to specify the relative speed, in percentages:
-    i.e., if ScaleUpPercent = 150 , then we can add 150% more pods (10 -> 25 pods)
-- `pods` - a parameter to specify the absolute speed, in the number of pods:
-    i.e., if ScaleUpPods = 5 , then we can add 5 more pods (10 -> 15)
-- `delay` - a parameter to specify the window over which the max (or min) recommendation is used. It behaves the same as the [Stabilization Window][].
+- `stabilizationWindowSeconds` - this value indicates the amount of time the HPA controller should consider
+  previous recommendations to prevent flapping of the number of replicas.
+- `scaleUp` specifies the rules which are used to control scaling behavior while scaling up
+  - `periodSecond` the amount of time in seconds for which the rule should hold true.
+  - `maxAllowedChanged` the maximum allowed changed in replicas in the given period. Can be an absolute
```

Comment:
I find it confusing to use a single field for both absolute and percentage values.  How will you tell the difference?  Is 0.5 a percentage or absolute value?

What about something more along the lines of @thockin's [suggestion](https://github.com/kubernetes/enhancements/pull/883#discussion_r280170553) of having "Pods" and "Percent" policies?

```yaml
constraints:
  scaleUp: # allow the max of 3 pods or 100% to be added every 15 seconds
  - policy: Pods
    value: 3
    periodSeconds: 15
  - policy: Percent
    value: 100
    periodSeconds: 15
  scaleDown: # allow the max of 2 pods or 20% to be removed every 15 seconds
  - policy: Pods
    value: 2
    periodSeconds: 15
  - policy: Percent
    value: 20
    periodSeconds: 15
  stabilizationWindowSeconds: 120 # downscale stabilization window of 2 minutes
```

Note: I changed the `scaleDown` field to be a list of policies so you can specify both Pods and Percent.  This would be useful for:
1. getting the autoscaler off the ground (i.e. allowing increase of 1->4, but still limiting to 200%)
2. allowing for future policy (I don't know what).

Reply:
I thought about this and the reason I went with a single value is because there is precedent for this as suggested by @liggitt [here](https://github.com/kubernetes/enhancements/pull/883#discussion_r264481727). Also the [deployment controller](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#rolling-update-deployment) currently uses the same type of value to specify increase in number of pods. Also there are already many hidden levers in the HPA controller which makes it harder for end-users to predict or analyze HPA behavior. Having a single value reduces the thinking you need to do when there are multiple policies.
 
Having multiple policies would make the HPA more flexible but at the moment I don't see the real need for it other than the use case of going from 1 -> 4 pods. Nor do I know if any type of workload which would need behavior like this. Also since we are going to split the code paths I don't think there is a need to preserve the old behavior, especially since it's not documented. As for a value like _"0.5"_ we could look into how it is treated in the deployment controller. 

That said if the maintainers think that multiple policies is the way to go then I will modify the proposal.

Reply:
There are examples of policy lists in the autoscaling space, such as [MetricSource](https://github.com/kubernetes/kubernetes/blob/8126201b73b3fa52bd376fc98cb00637af493b72/staging/src/k8s.io/api/autoscaling/v2beta2/types.go#L74-L118).

Given that this will live in the same resource (HPA) as MetricTarget, I would prefer to follow that pattern over the Deployment pattern.  I see your point about a single value being easier to understand, but I would prefer flexibility in this case.

> I don't see the real need for it

A hypothetical example: what if there were a new policy which limits the core-hours used per day?  A sort of "cost" constraint.

```yaml
constraints:
  scaleUp:
  - policy: Cost
    cpuCoresHours: 240    # 10 cores per day on average
    periodSeconds: 86400  # budgeted over a 24 hour window
```

We don't have to design it now.  But I would like to avoid an autoscaling v2beta3 by leaving space for it.  :crossed_fingers: